### PR TITLE
Edge case when transferring blocks

### PR DESCRIPTION
**Pull request by abravalheri**
This is actually an issue disguised as a PR (I thought it was easier if I provided the test case that is failing and let the CI show it fails).

The issue is as follows:

- All the blocks in the ConfigUpdater hierarchy have a `container` reference, but when you move the block around that reference not necessarily is removed (the types are safe for the time being because `block._container: Container[T]`, so in no moment the current implementation considers the possibility of `block._container` being `None` or even add checks for it).
- This reference is used in the block builder API
- If a node is removed from its parent, or it is imported into a different parent, then there might be inconsistencies, and the block builder API might do something unexpected.

The example implemented in the test case shows section nodes being parsed from text and added to a different document.
When we import a section from a different ConfigUpdater object, its internal reference to its container does not change.
So if we use the `add_after` method to add a new block, this new block will be added to the wrong container.

This happens even if we use `copy` and even if we get a completely new reference to the `section` object after it is added, using the correct parent.

### CODE REVIEW

File: tests/test_transfer_blocks.py [hunk 1]
```diff
@@ -0,0 +1,46 @@
+from copy import copy
+from textwrap import dedent
+
+from configupdater import ConfigUpdater
+
+
+def test_transfering_blocks_between_elments():
```
Comment (FlorianWilhelm) [inline tests/test_transfer_blocks.py:7]: There is an `e` missing in elements.

**Comment by FlorianWilhelm** [comment]
Interesting, this is definitely a bug and the reference to the container needs to be set correctly when inserted. I haven't had a close look but I guess `__setitem__` should make sure of that.

**Comment by abravalheri** [comment]
@FlorianWilhelm how do you feel about delaying the assignment, instead of doing it in the moment the class is created, do it later when it is inserted in the structure? I can see some benefits for this approach (like being able to have a `__repr__` that is actually executable Python code and that can be used for debugging, like the `dataclasses` implementation.

I could give it a shot.

**Comment by FlorianWilhelm** [comment]
> @FlorianWilhelm how do you feel about delaying the assignment, instead of doing it in the moment the class is created, do it later when it is inserted in the structure? I can see some benefits for this approach (like being able to have a `__repr__` that is actually executable Python code and that can be used for debugging, like the `dataclasses` implementation.
> 
> I could give it a shot.

Hm... seems like a good idea when users want to work with `Option`, `Comment`, etc. directly. Do you think they will do? I thought that only `ConfigUpdater` itself will create those instances and so I decided to pass the container directly at creation time. 

> 
> Regarding this issue, I think ideally users should call `.remove()` before attaching blocks to a new container. The `copy` pattern does not help either... Nodes should not pass their "attachment state" to their copies, other wise we can copy a block, remove the original, but still access the container from the copy...

How about calling `copy` within `__setitem__` and there also setting the container correctly? Then the source section (from your unit test) stays unmodified and the edge case in `target` should be gone, right?

**Comment by abravalheri** [comment]
> Hm... seems like a good idea when users want to work with Option, Comment, etc. directly. Do you think they will do? I thought that only ConfigUpdater itself will create those instances and so I decided to pass the container directly at creation time.

The thing that I like the most about this approach is being able to inspect things going wrong in the terminal and be able to copy and paste the output and run it again:
```python
>>> print(repr(updater['section']))
Section("name", [Option("a", "value"), Option("b", "value"), Comment("# comment")])
```
The `dataclasses` and Pydantic follow this approach and it is very nice to work with specially in an interactive session.

---

> How about calling copy within __setitem__ and there also setting the container correctly? Then the source section (from your unit test) stays unmodified and the edge case in target should be gone, right?

That should work too, but when calling `.remove()` we should also invalidate the reference somehow...

**Comment by abravalheri** [comment]
In an unrelated note: `weakref`s are a good call for this use case of cyclic dependencies too

**Comment by abravalheri** [comment]
Just thinking here, that copying solve the problem if it is not nested, i.e. I copy a section and call the builder API in the copied section.

If I copy a section, but then call the builder API in an option inside the copied section, there will be 2 problems:

- the options inside the copied section still have a reference to the original section
- the original section and the copied section share a shallow copy of the `_structure`, so If I change the section contents in the source object, the target object will see the same changed `_structure`.

**Comment by FlorianWilhelm** [comment]
Hm... let me think about this. Don't know when I find the time but I think this could be nicely done by overloading `__deepcopy__`. 

Regarding the `remove`, I don't understand why the user should have to care about this? We could copy by default and set the right reference to the new container by overwriting the old one. Maybe I am missing something obvious here ;-)
So from the unit test in PR #35:
```python
updater = ConfigUpdater()
sect_updater = ConfigUpdater()
updater.read_string(test6_cfg_in)
sect_updater.read_string(test6_cfg_in)

section = sect_updater["section0"]
updater["new_section"] = section.remove()
```

Why could not `__setitem__` do the `remove()` or just overwrite with the new reference after having correctly copied `section`? Then a user could never forget to call `remove`. The only difference is that `section` itself will stay independent of `updater` which I think is a good thing. It should be mentioned in the docs.

**Comment by abravalheri** [comment]
Hi @FlorianWilhelm thanks for the review.

Regarding you question about `remove`, I am not sure myself if it is a good idea to send a warn asking for remove to be called explicitly. In fact my first implementation in #35 did not do that. Afterwards I came back and added the warning.

The only thing `remove` brings is explicitness. I am more than happy to drop the warning in #35 and revert the test case to not use `remove`.

Well, there is one extra thing about the `remove`... the problem with the nested references is also solved with `remove` being required to called explicitly, because them the nested references are always right and no objects share `_structure`...

**Comment by abravalheri** [comment]
Thank you very much @FlorianWilhelm. The errors when importing section should be covered in #39, right?
Do we have also to be concerned about options being moved from one document to the other, or that would also be covered by #39?

**Comment by abravalheri** [comment]
Thank you very much @FlorianWilhelm.

This should be covered in #39 and #44
